<?xml version="1.0"?>
<launch>

    <!-- Arm configuration -->
    <arg name="debug" default="false"/>
    <arg name="arm" default="gen3"/>
    <arg name="dof" default="7"/>
    <arg name="vision" default="true"/> <!-- True if the arm has a Vision module -->
    <arg name="use_hard_limits" default="false"/> <!-- Set soft limits as hard limits (Gen3 only) -->
  
    <!-- Gripper configuration -->
    <arg name="gripper" default=""/>
    <arg name="has_gripper" default="true" if="$(eval arg('gripper')=='robotiq_2f_85')"/>
    <arg name="has_gripper" default="false" if="$(eval arg('gripper')!='robotiq_2f_85')"/>
    
    <!-- Sensors -->
    <arg name="zedm" default="false"/>
    <arg name="tool_frame_offset" default="0.12"/>

    <!-- Namespace -->
    <arg name="robot_name" default="my_$(arg arm)"/>
    <arg name="prefix" default=""/>

    <!-- Payload options (only supported without gripper) -->
    <arg name="add_payload" default="false"/>
    <arg name="payload_mass" default="1.2522"/>
    <arg name="payload_offset" default="0.087"/>

    <!-- Kortex API options -->
    <arg name="ip_address" default="192.168.1.10"/>
    <arg name="username" default="admin"/>
    <arg name="password" default="admin"/>
    <arg name="cyclic_data_publish_rate" default="40"/> <!--Hz-->
    <arg name="api_rpc_timeout_ms" default="2000"/> <!--milliseconds-->
    <arg name="api_session_inactivity_timeout_ms" default="35000"/> <!--milliseconds-->
    <arg name="api_connection_inactivity_timeout_ms" default="20000"/> <!--milliseconds-->

    <!-- Automatically start other modules -->
    <arg name="start_rviz" default="false"/>
    <arg name="start_moveit" default="true"/>

    <!-- Action server params -->
    <arg name="default_goal_time_tolerance" default="0.5"/> <!--seconds-->
    <arg name="default_goal_tolerance" default="0.005"/> <!--radians-->

    <group ns="$(arg robot_name)">
        <group if="$(eval not arg('prefix'))">

            <!-- Load the description for the robot -->
            <!-- Without gripper -->
            <param name="robot_description" command="$(find xacro)/xacro --inorder $(find kortex_description)/robots/$(arg arm).xacro dof:=$(arg dof) vision:=$(arg vision) sim:=false
                 tool_frame_offset:=$(arg tool_frame_offset) add_payload:=$(arg add_payload) payload_mass:=$(arg payload_mass) payload_offset:=$(arg payload_offset)" 
                   unless="$(eval arg('has_gripper'))"/>
            
            <!-- With gripper -->
            <param name="robot_description" command="$(find xacro)/xacro --inorder $(find kortex_description)/robots/$(arg arm)_$(arg gripper).xacro dof:=$(arg dof) vision:=$(arg vision) sim:=false tool_frame_offset:=$(arg tool_frame_offset)"
                   if="$(eval arg('has_gripper'))"/>
        </group>
        <group unless="$(eval not arg('prefix'))">
            <!-- Load the description for the robot -->
            <!-- Without gripper -->
            <param name="robot_description" command="$(find xacro)/xacro --inorder $(find kortex_description)/robots/$(arg arm).xacro dof:=$(arg dof) vision:=$(arg vision) sim:=false prefix:=$(arg prefix) tool_frame_offset:=$(arg tool_frame_offset)  add_payload:=$(arg add_payload) payload_mass:=$(arg payload_mass) payload_offset:=$(arg payload_offset)"
                   unless="$(eval arg('has_gripper'))"/>
            <!-- With gripper -->
            <param name="robot_description" command="$(find xacro)/xacro --inorder $(find kortex_description)/robots/$(arg arm)_$(arg gripper).xacro dof:=$(arg dof) vision:=$(arg vision) sim:=false prefix:=$(arg prefix) tool_frame_offset:=$(arg tool_frame_offset)"
                   if="$(eval arg('has_gripper'))"/>
        </group>

        <!-- Manually specified joint limits, zeros and low level pd gains -->
        <rosparam command="load" file="$(find kinova_robot)/config/joint_limits.yaml" subst_value="true"/>
        <rosparam command="load" file="$(find kinova_robot)/config/arm_pid.yaml" subst_value="true"/>

        <!-- Start the kortex_driver node -->
        <!-- <arg name="launch_prefix" default="valgrind /-/-leak-check=full /-/-track-origins=yes -v" if="$(arg debug)"/> -->
        <arg name="launch_prefix" default="gdb -ex run --args" if="$(arg debug)"/>
        <arg name="launch_prefix" default="" unless="$(arg debug)"/>
        <node name="kinova_robot" pkg="kinova_robot" type="kinova_robot_node" output="screen" launch-prefix="$(arg launch_prefix)">
            <param name="sim" value="false"/>
            <param name="ip_address" value="$(arg ip_address)"/>
            <param name="username" value="$(arg username)"/>
            <param name="password" value="$(arg password)"/>
            <param name="cyclic_data_publish_rate" value="$(arg cyclic_data_publish_rate)"/>
            <param name="api_rpc_timeout_ms" value="$(arg api_rpc_timeout_ms)"/>
            <param name="api_session_inactivity_timeout_ms" value="$(arg api_session_inactivity_timeout_ms)"/>
            <param name="api_connection_inactivity_timeout_ms" value="$(arg api_connection_inactivity_timeout_ms)"/>
            <param name="default_goal_time_tolerance" value="$(arg default_goal_time_tolerance)"/>
            <param name="default_goal_tolerance" value="$(arg default_goal_tolerance)"/>
            <param name="arm" value="$(arg arm)"/>
            <param name="gripper" value="$(arg gripper)"/>
            <param name="dof" value="$(arg dof)"/>
            <param name="use_hard_limits" value="$(arg use_hard_limits)"/>
            <param name="robot_name" value="$(arg robot_name)"/>
            <param name="prefix" value="$(arg prefix)"/>
            <rosparam command="load" file="$(find kortex_description)/arms/$(arg arm)/$(arg dof)dof/config/joint_limits.yaml" subst_value="true"/>

            <!-- If there is a gripper, load the active joint names for it -->
            <rosparam command="load" file="$(find kortex_description)/grippers/$(arg gripper)/config/joint_limits.yaml" unless="$(eval not arg('has_gripper'))" subst_value="true"/>

            <!-- Joint zeros (calibration params) -->
            <rosparam command="load" file="$(find kinova_robot)/config/joint_zeros.yaml" subst_value="true"/>
            <param name="calibration_file" value="$(find kinova_robot)/config/joints.yaml"/>
        </node>

        <!-- Start MoveIt! main executable -->
        <group if="$(arg start_moveit)">

            <!-- TODO Find cleaner way to do that and that will work with other arms -->
            <!-- Without gripper -->
            <include file="$(find kortex_description)/../kortex_move_it_config/$(arg arm)_move_it_config/launch/move_group.launch" if="$(eval not arg('has_gripper'))">
                <arg name="dof" value="$(arg dof)"/>
                <arg name="use_hard_limits" value="$(arg use_hard_limits)"/>
                <arg name="prefix" value="$(arg prefix)"/>
            </include>
            <!-- With gripper -->
            <include file="$(find kortex_description)/../kortex_move_it_config/$(arg arm)_$(arg gripper)_move_it_config/launch/move_group.launch" unless="$(eval not arg('has_gripper'))">
                <arg name="dof" value="$(arg dof)"/>
                <arg name="use_hard_limits" value="$(arg use_hard_limits)"/>
                <arg name="prefix" value="$(arg prefix)"/>
            </include>

        </group>

        <!-- Start joint and robot state publisher -->
        <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
            <rosparam param="source_list">[base_feedback/joint_state]</rosparam>
            <param name="use_gui" value="false"/>
            <param name="rate" value="$(arg cyclic_data_publish_rate)" />
        </node>
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />

        <!-- Start RViz -->
        <node name="rviz" pkg="rviz" type="rviz" output="log" args="-d $(find kortex_driver)/config.rviz" if="$(arg start_rviz)"/>

        <!-- Tests -->
        <!-- Initialization test and custom functional tests -->
        <test test-name="kortex_driver_tests" pkg="kortex_driver" type="kortex_arm_driver_func_tests">
            <param name="ip_address" value="$(arg ip_address)"/>
            <param name="cyclic_data_publish_rate" value="$(arg cyclic_data_publish_rate)"/>
            <param name="api_rpc_timeout_ms" value="$(arg api_rpc_timeout_ms)"/>
            <param name="api_session_inactivity_timeout_ms" value="$(arg api_session_inactivity_timeout_ms)"/>
            <param name="api_connection_inactivity_timeout_ms" value="$(arg api_connection_inactivity_timeout_ms)"/>
            <param name="default_goal_time_tolerance" value="$(arg default_goal_time_tolerance)"/>
            <param name="default_goal_tolerance" value="$(arg default_goal_tolerance)"/>
            <param name="arm" value="$(arg arm)"/>
            <param name="gripper" value="$(arg gripper)"/>
            <param name="dof" value="$(arg dof)"/>
            <rosparam command="load" file="$(find kortex_description)/arms/$(arg arm)/$(arg dof)dof/config/joint_limits.yaml" subst_value="true"/>
            <!-- If there is a gripper, load the active joint names for it -->
            <rosparam command="load" file="$(find kortex_description)/grippers/$(arg gripper)/config/joint_limits.yaml" unless="$(eval not arg('has_gripper'))" subst_value="true"/>
        </test>

        <!-- Test feedback and joint state publishing -->
        <test test-name="publish_test_kortex_driver" pkg="rostest" type="publishtest">
            <rosparam subst_value="true">
                topics:
                    - name: /$(arg robot_name)/base_feedback
                      timeout: 10
                      negative: False
                    - name: /$(arg robot_name)/base_feedback/joint_state
                      timeout: 10
                      negative: False
                    - name: /$(arg robot_name)/joint_states
                      timeout: 10
                      negative: False
            </rosparam>
        </test>

        <!-- Test publishers rate -->
        <test test-name="hztest_test_base_feedback" pkg="rostest" type="hztest">
            <param name="topic" value="base_feedback" />
            <param name="hz" value="$(arg cyclic_data_publish_rate)" />
            <param name="hzerror" value="5.0" />
            <param name="test_duration" value="30.0" />
            <param name="wait_time" value="10.0" />
        </test>
        <test test-name="hztest_test_driver_joint_state" pkg="rostest" type="hztest">
            <param name="topic" value="base_feedback/joint_state" />
            <param name="hz" value="$(arg cyclic_data_publish_rate)" />
            <param name="hzerror" value="5.0" />
            <param name="test_duration" value="30.0" />
            <param name="wait_time" value="10.0" />
        </test>
        <test test-name="hztest_test_joint_state_publisher_joint_states" pkg="rostest" type="hztest">
            <param name="topic" value="joint_states" />
            <param name="hz" value="$(arg cyclic_data_publish_rate)" />
            <param name="hzerror" value="5.0" />
            <param name="test_duration" value="30.0" />
            <param name="wait_time" value="10.0" />
        </test>

    </group>

    <group if="$(arg zedm)">
      <include file="$(find kinova_sensors)/launch/load_zed.launch"/>
      
    </group>
</launch>
